commit 76ae751c7caf1f881932a20d93fe56c4fd3cd743 (refs/changes/51/3051/1)
Author: Rob Johnston <rob.johnston@joyent.com>
Date:   2017-12-06T00:11:54+00:00 (1 year, 10 months ago)
    
    OS-6478 default chunk size used by ipmi_fru_read() is too large for some SP's

diff --git a/usr/src/lib/libipmi/common/ipmi_fru.c b/usr/src/lib/libipmi/common/ipmi_fru.c
index d652c23cd4..4796cd0d01 100644
--- a/usr/src/lib/libipmi/common/ipmi_fru.c
+++ b/usr/src/lib/libipmi/common/ipmi_fru.c
@@ -22,9 +22,9 @@
  * Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
  * Use is subject to license terms.
  */
-
-#pragma ident	"%Z%%M%	%I%	%E% SMI"
-
+/*
+ * Copyright (c) 2017, Joyent, Inc.
+ */
 #include <libipmi.h>
 #include <string.h>
 
@@ -52,7 +52,8 @@ int
 ipmi_fru_read(ipmi_handle_t *ihp, ipmi_sdr_fru_locator_t *fru_loc, char **buf)
 {
 	ipmi_cmd_t cmd, *resp;
-	uint8_t count, devid;
+	int ierrno;
+	uint8_t count, devid, chunksz;
 	uint16_t sz, offset = 0;
 	ipmi_fru_read_t cmd_data_in;
 	char *tmp;
@@ -82,14 +83,15 @@ ipmi_fru_read(ipmi_handle_t *ihp, ipmi_sdr_fru_locator_t *fru_loc, char **buf)
 		return (-1);
 	}
 
+	chunksz = 128;
 	while (offset < sz) {
 		cmd_data_in.ifr_devid = devid;
 		cmd_data_in.ifr_offset_lsb = BITX(offset, 7, 0);
 		cmd_data_in.ifr_offset_msb = BITX(offset, 15, 8);
-		if ((sz - offset) < 128)
+		if ((sz - offset) < chunksz)
 			cmd_data_in.ifr_count = sz - offset;
 		else
-			cmd_data_in.ifr_count = 128;
+			cmd_data_in.ifr_count = chunksz;
 
 		cmd.ic_netfn = IPMI_NETFN_STORAGE;
 		cmd.ic_cmd = IPMI_CMD_READ_FRU_DATA;
@@ -98,6 +100,13 @@ ipmi_fru_read(ipmi_handle_t *ihp, ipmi_sdr_fru_locator_t *fru_loc, char **buf)
 		cmd.ic_lun = 0;
 
 		if ((resp = ipmi_send(ihp, &cmd)) == NULL) {
+			ierrno = ipmi_errno(ihp);
+			if (chunksz > 16 &&
+			    (ierrno == EIPMI_DATA_LENGTH_EXCEEDED ||
+			    ierrno == EIPMI_INVALID_REQUEST)) {
+				chunksz = chunksz >> 1;
+				continue;
+			}
 			free(tmp);
 			return (-1);
 		}
